### ISSUE TITLE
SVG Fails to display properly

### ISSUE THREAD

**Issue by augyg** [obsidiansystems/obelisk#1064]
Fail State:
![image](https://github.com/obsidiansystems/obelisk/assets/36346907/a0660705-4c83-4a37-8dea-42b909c81c25)

Should look like: 
![image](https://github.com/obsidiansystems/obelisk/assets/36346907/da1055d9-2245-4d90-9cda-9c85d3afc507)

given the code:
![image](https://github.com/obsidiansystems/obelisk/assets/36346907/5ef392ef-892f-46b9-91d5-e8b53ecea7be)

Note: the comment in the code made no difference with or without

This happens on Firefox and Chrome ( which is all that I've tested )

this page is deployed here, to test out the behavior I describe next ( https://aceinterviewprep.app/landing/base )

On initial request to get the page it renders perfectly fine however if you click on any link (eg. login) and then hit the back button or click solutions > coding bootcamps (or any) then click the Ace logo to go back then you will see the top image 

I have no understanding of why this is happening but even if my code is funky, it should still render the same each time right?

**Comment by augyg** [comment]
I can see that when it loads properly it has the **data-ssr attribute**

I have no idea why or if this matters though, I'm guessing hydration?

![image](https://github.com/obsidiansystems/obelisk/assets/36346907/bd7c0703-82ea-497f-a764-068a287eb1f3)

**Comment by augyg** [comment]
It seems that the only way it displays properly is from the initial server side render. 

prerender_ mySvg mySvg also recreates the issue

**Comment by alexfmpe** [comment]
https://github.com/obsidiansystems/obelisk/issues/828 seems relevant
Think I also ran into this a long time ago, but workarounded some other way.

**Comment by augyg** [comment]
thanks @alexfmpe, based on what I saw with the outdated reflex-dom-SVG package and some investigation, I wouldn't be surprised if the workaround in #828 would work here

I just made the SVG in figma instead but hopefully I can test this soon. 

Feels like this should become part of FAQ for reflex-dom as the behavior is relatively surprising, I wonder if reflex-dom-contrib has solutions yet

**Comment by tdimiduk** [comment]
For reasons I don't completely understand, you have to set the namespace at a lower level in the element api. For example I used this function:

```
svgXMLNamespace :: Text
svgXMLNamespace = "http://www.w3.org/2000/svg"

elAttrNS' ::
  DomBuilder t m =>
  Maybe Namespace ->
  Text ->
  Map.Map Text Text ->
  m a ->
  m (Element EventResult (DomBuilderSpace m) t, a)
elAttrNS' mns elementTag attrs =
  element elementTag $
    def
      & elementConfig_namespace .~ mns
      & initialAttributes .~ Map.mapKeys (AttributeName Nothing) attrs`
```
Or [elDynAttrNS](https://github.com/reflex-frp/reflex-dom/blob/develop/reflex-dom-core/src/Reflex/Dom/Widget/Basic.hs#L237) does it correctly.

the key bit is that you have to set this specific attr with `elementConfig_namespace .~ "http://www.w3.org/2000/svg"` it doesn't work setting it at the normal attrs level. It's probably worth documenting this somewhere, I'd be happy to write something but I'm not sure where is the appropriate place to put it.

**Comment by alexfmpe** [comment]
> the key bit is that you have to set this specific attr with elementConfig_namespace .~ "http://www.w3.org/2000/svg" it doesn't work setting it at the normal attrs level. It's probably worth documenting this somewhere, I'd be happy to write something but I'm not sure where is the appropriate place to put it.

Maybe it's a bug with hydration? 
After a quick glance nothing stands out to me with the usage of `data-ssr` but it's strange `_elementConfig_namespace` is treated specially. 
Anything come to mind @tomsmalley ?

**Comment by tomsmalley** [comment]
> > the key bit is that you have to set this specific attr with elementConfig_namespace .~ "http://www.w3.org/2000/svg" it doesn't work setting it at the normal attrs level. It's probably worth documenting this somewhere, I'd be happy to write something but I'm not sure where is the appropriate place to put it.
> 
> Maybe it's a bug with hydration? After a quick glance nothing stands out to me with the usage of `data-ssr` but it's strange `_elementConfig_namespace` is treated specially. Anything come to mind [@tomsmalley](https://github.com/tomsmalley) ?

Really testing my long term memory here Alex 😄 

The namespace config entry is the key one for making the immediate builder use the correct createElement function:

```haskell
  e <- uncheckedCastTo DOM.Element <$> case cfg ^. namespace of
    Nothing -> createElement doc elementTag
    Just ens -> createElementNS doc (Just ens) elementTag
```
https://github.com/reflex-frp/reflex-dom/blob/f975e57044bfa8020ee7bfcb189203559adf3f31/reflex-dom-core/src/Reflex/Dom/Builder/Immediate.hs#L708-L710

But the only circumstances you should hit this are:
1) if hydration has failed and the element is being created dynamically (would be concerning, should result in console errors iirc), or
2) if you're in the right hand side of `prerender`, which was mentioned above.

_Not_ using the NS variants of the element functions will lead to correct rendering on initial load (provided SSR is working - browser just sees the namespace attr in the html and acts accordingly), but any dynamic element creation will be wrong.

TLDR, you need to use the NS variants if you need a namespaced element.

(You might also be able to get away with a `skipHydrationAttribute` on the left hand side of `prerender` too, with `pure ()` on the right hand side. Then it'll be rendered serverside, the browser sees the correct HTML and the SVG works, and then hydration skips it. That's a bit of guesswork though.)

**Comment by tomsmalley** [comment]
Oh, I just read the original issue a bit more closely.

> On initial request to get the page it renders perfectly fine however if you click on any link (eg. login) and then hit the back button or click solutions > coding bootcamps (or any) then click the Ace logo to go back then you will see the top image

So hydration is working, but when you switch pages (and you're using the obelisk router stuff to prevent a full page load) the element will be recreated dynamically. Which will not work, because of what I said above. Use the NS variant instead and it should work in both cases.

**Comment by tdimiduk** [comment]
@tomsmalley Thanks for the context! The immediate builder needing to pick the correct `createElement` function is the bit of context I was missing to understand why this needs the different lower level element creation functions. 

Any thoughts on where we might want to write a note about this nuance? Should we call out svg specifically in the docs for the NS variant? is it worth having an SVG specific function somewhere? Does this deserve a mention in the obelisk or reflex-dom faqs? Is somebody bringing https://hackage.haskell.org/package/reflex-dom-svg up to date or writing a new package where this kind of thing belongs?

**Comment by tomsmalley** [comment]
I suppose calling out the difference between NS and non-NS should be done here:

https://hackage.haskell.org/package/reflex-dom-core-0.8.1.2/docs/Reflex-Dom-Widget-Basic.html#g:2

It looks like there are some NS variants missing too, which might be worth adding.
